### PR TITLE
Reproduce Float-passing problem

### PR DESCRIPTION
**Pull request by GregoryTravis**
This applies @JaroslavTulach's fix in the comment below.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: test/Base_Tests/src/Data/Decimal_Spec.enso [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                + [-92233720.0]
+                + [-9223372.0]
+                + [-922337.0]
+                + [-92233.0]
+                + [-9223.0]
+                + [-922.0]
+
+            vs.map v->
+                r = float_id v
+                IO.println "float_id "+v.to_text+" "+r.to_text
```
Comment (GregoryTravis) [inline test/Base_Tests/src/Data/Decimal_Spec.enso]: @JaroslavTulach I am seeing some unusual behavior here. I wonder if you get the same results when running `Decimal_Spec.enso` locally?

This test simply passes a Float to a Java function that returns its argument. It seems like the argument is changed when it reaches the Java in some cases.

```
float_id 9.223372036854776E19 9.223372036854776E19
float_id 9.223372036854778E18 9.223372036854778E18
float_id 9.223372036854776E18 9.223372036854776E18
float_id 9.223372036854776E18 9.223372036854776E18
float_id 9.223372036854776E18 9.223372036854776E18
float_id 9.223372036854775E18 NaN # unexpected
float_id 9.223372036854774E18 NaN # unexpected
float_id 9.2233720368547E18 NaN # unexpected
float_id 9.223372036854E18 NaN # unexpected
float_id 9.223372E18 NaN # unexpected
float_id 9.2233720368547E17 5.148800401705924E-247 # unexpected
float_id 9.2233720368547E16 1.7265376942307606E-302 # unexpected
float_id 9.2233720368547E15 4.663754807431019E-308 # unexpected
float_id 9.2233720368547E14 9.2233720368547E14
float_id 9.2233720368547E13 9.2233720368547E13
float_id 9.223372036854E12 9.223372036854E12
float_id 9.22337203685E11 9.22337203685E11
float_id 9.2233720368E10 9.2233720368E10
float_id 9.223372036E9 9.223372036E9
float_id 9.22337203E8 9.22337203E8
float_id 9.223372E7 9.223372E7
float_id 9223372.0 9223372.0
float_id 922337.0 922337.0
float_id 92233.0 92233.0
float_id 9223.0 9223.0
float_id 922.0 922.0
float_id -9.223372036854776E19 -9.223372036854776E19
float_id -9.223372036854778E18 -9.223372036854778E18
float_id -9.223372036854776E18 -0.0 # unexpected
float_id -9.223372036854776E18 -0.0 # unexpected
float_id -9.223372036854776E18 -0.0 # unexpected
float_id -9.223372036854775E18 -5.06E-321 # unexpected
float_id -9.223372036854774E18 -1.012E-320 # unexpected
float_id -9.2233720368547E18 -3.74383E-319 # unexpected
float_id -9.223372036854E18 -3.8349E-318 # unexpected
float_id -9.223372E18 -1.8208678612E-313 # unexpected
float_id -9.2233720368547E17 -8.390303882377555E246 # unexpected
float_id -9.2233720368547E16 -2.6059089326772877E302 # unexpected
float_id -9.2233720368547E15 -8.772742498128177E307 # unexpected
float_id -9.2233720368547E14 -9.2233720368547E14
float_id -9.2233720368547E13 -9.2233720368547E13
float_id -9.223372036854E12 -9.223372036854E12
float_id -9.22337203685E11 -9.22337203685E11
float_id -9.2233720368E10 -9.2233720368E10
float_id -9.223372036E9 -9.223372036E9
float_id -9.22337203E8 -9.22337203E8
float_id -9.223372E7 -9.223372E7
float_id -9223372.0 -9223372.0
float_id -922337.0 -922337.0
float_id -92233.0 -92233.0
float_id -9223.0 -9223.0
float_id -922.0 -922.0
```

Comment (JaroslavTulach) [reply test/Base_Tests/src/Data/Decimal_Spec.enso]: > I wonder if you get the same results when running `Decimal_Spec.enso` locally?

Yes, I do. I am seeing the same results.

**Comment by JaroslavTulach** [comment]
This change seems to fix the problem:
```diff
enso$ git diff engine/polyglot-api/src/main/java/org/enso/polyglot/HostAccessFactory.java
diff --git engine/polyglot-api/src/main/java/org/enso/polyglot/HostAccessFactory.java engine/polyglot-api/src/main/java/org/enso/polyglot/HostAccessFactory.java
index 1f9cbbcf2c..150c43fc2f 100644
--- engine/polyglot-api/src/main/java/org/enso/polyglot/HostAccessFactory.java
+++ engine/polyglot-api/src/main/java/org/enso/polyglot/HostAccessFactory.java
@@ -21,11 +21,6 @@ public class HostAccessFactory {
         .allowIteratorAccess(true)
         .allowMapAccess(true)
         .allowAccessInheritance(true)
-        .targetTypeMapping(
-            Long.class,
-            Double.class,
-            (v) -> v != null && !longInSafeDoubleRange(v, LONG_MAX_SAFE_DOUBLE),
-            (v) -> Double.longBitsToDouble(v))
         .build();
   }
 
```
This conversion has been added in
- #4099

by @hubertp. Hubert, looks like @GregoryTravis found edge cases when the conversion isn't working properly.

**Comment by JaroslavTulach** [comment]
We seem to have [a failure](https://github.com/enso-org/enso/actions/runs/8102254669/job/22144312365?pr=9201#step:8:4577) related to **Numeric values: can be passed in host calls without lossy coercion exception** as that now ends:
```
Reason: An unexpected panic was thrown: (Unsupported_Argument_Types.Error [6907338656278321365] 
Cannot convert '6907338656278321365'(language: Java, type: java.lang.Long) to Java type 'double': 
Invalid or lossy primitive coercion.)
 at <enso> <anonymous><arg-1>(test/Base_Tests/src/Semantic/Java_Interop_Spec.enso:53:13-34)
 at <enso> null(Internal)
 at <enso> <anonymous>(Standard/Test/2024.1.1-dev/src/Group.enso:25:61-64)
```
I assume this is the test added by #4099 - its implementation was achieved by those [five deleted lines](https://github.com/enso-org/enso/pull/9201/commits/e04fcf191f5745de58eee913b971017727709e61#diff-177df2cde25ff8be857083cfdf31b23ff2e32ac3f797b23ac9b2b75e2f592498L24). Now we have two options:

- remove the test
- don't delete [the lines](https://github.com/enso-org/enso/pull/9201/commits/e04fcf191f5745de58eee913b971017727709e61#diff-177df2cde25ff8be857083cfdf31b23ff2e32ac3f797b23ac9b2b75e2f592498L24), but replace them with better ones

Removing the test can probably be done by anyone (with the blessing by someone from _libraries team_). Finding better lines is trickier. @hubertp please take a look and propose an alternative that wouldn't have drawbacks demonstrated in this PR and still satisfy the test.

**Comment by JaroslavTulach** [comment]
> * remove the test

Removing a test is a bad practice in general, but it may be **the right solution in this case**. I believe the test has been written before support for 
- #7213 

has been put in place. Before the `BigInteger` interop got fully supported there was no way to get `java.math.BigInteger` into libraries Java code - as such the conversion to `double` might have been necessary. However now it is possible to accept `java.math.BigInteger` along `double` or `long` and handle the case differently.

If that's the case the test may be deleted or modified to benefit from the new _big integer capabilities_.

**Comment by radeusgd** [comment]
> > * remove the test
> 
> Removing a test is a bad practice in general, but it may be **the right solution in this case**. I believe the test has been written before support for
> 
> * [EnsoBigInteger should be a BigInteger polyglot type #7213](https://github.com/enso-org/enso/issues/7213)
> 
> has been put in place. Before the `BigInteger` interop got fully supported there was no way to get `java.math.BigInteger` into libraries Java code - as such the conversion to `double` might have been necessary. However now it is possible to accept `java.math.BigInteger` along `double` or `long` and handle the case differently.
> 
> If that's the case the test may be deleted or modified to benefit from the new _big integer capabilities_.

I agree that once we have proper `BigInteger` support, we can modify our methods to accept a `BigInteger` properly.

But I don't think that was the only motivation behind #4099.

1. One thing is APIs that we control (e.g. `Moments` in the test there), where we could add the `BigInteger` overload (and we probably also need a `long` overload there as well, right?)
2. Another thing is external APIs that we _do not control_, e.g. someone calling a Java function from _some_ library that is only accepting `double`.

The problem is - in Enso we usually accept `2` wherever a `Float` would be accepted, and so the major question is - are we okay with not allowing that in polyglot calls? Shall we then add some kind of `to_float` conversion, that allows to convert a `2` into `2.0`?

**Comment by GregoryTravis** [comment]
@hubertp What are your thoughts?

**Comment by hubertp** [comment]
> @hubertp What are your thoughts?

Sorry, I replied to @JaroslavTulach in DM. I'm fine with ditching the test. The patch has been added prior to BigInteger support. I would also be fine in this case to make conversions explicit, rather than implicit, now.
The additional advantage of removing target mapping is that context initialization will be faster (that fragment actually took a non-negligible amount of time and could not be bypassed with other means).

**Comment by GregoryTravis** [comment]
> > @hubertp What are your thoughts?
> 
> Sorry, I replied to @JaroslavTulach in DM. I'm fine with ditching the test. The patch has been added prior to BigInteger support. I would also be fine in this case to make conversions explicit, rather than implicit, now. The additional advantage of removing target mapping is that context initialization will be faster (that fragment actually took a non-negligible amount of time and could not be bypassed with other means).

Should this PR be merged on its own, or does it need other changes with it?

**Comment by JaroslavTulach** [comment]
> Should this PR be merged on its own, or does it need other changes with it?

Once the CI is green and code owners are fine with the PR, it can be merged.

**Review by JaroslavTulach — APPROVED** [review]

File: std-bits/base/src/main/java/org/enso/base/numeric/Number_Utils.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package org.enso.base.numeric;
+
+public class Number_Utils {
+  // For testing only.
+  public static Double floatId(Double d) {
+    return d;
+  }
+}
```
Comment (radeusgd) [inline std-bits/base/src/main/java/org/enso/base/numeric/Number_Utils.java]: If this is only used for tests, it should be moved to `test/Base_Tests/polyglot-sources`. This is a Java lib specifically compiled for `Base_Tests`.

**Review by radeusgd — APPROVED** [review]
Looks good, but I'd like to move the helper to the dedicated test library instead.

btw. @JaroslavTulach curious note for you - as you can see, we have a separate JAR built for `Base_Tests` that includes some utils which are used for the tests but are not meant to be 'shipped' for users of `Base` library. Do you think we can retain this structure if we merge the tests to be parts of their libraries? (I guess it would not be a big issue if we cannot, just wondering)

File: std-bits/base/src/main/java/org/enso/base/numeric/Number_Utils.java (hunk 2 above)
Comment (GregoryTravis) [reply std-bits/base/src/main/java/org/enso/base/numeric/Number_Utils.java]: Done

**Review by hubertp — APPROVED** [review]

**Pull request merged**